nifty-ssl/src/main/java/com/facebook/nifty/ssl/PollingMultiFileWatcher.java

```diff
@@ -0,0 +1,275 @@
+import com.google.common.io.BaseEncoding;
+import com.google.common.io.Files;
+import com.google.common.util.concurrent.ListenableScheduledFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.airlift.log.Logger;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.FileTime;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * An implementation of {@link MultiFileWatcher} that polls the watched files for changes at regular intervals
+ * in a background thread.
+ */
+public final class PollingMultiFileWatcher implements MultiFileWatcher {
+    // Fields set in constructor
+    private final long initialDelay;
+    private final long interval;
+    private final TimeUnit timeUnit;
+    // Fields set in registerUpdateCallback()
+    private Consumer<Set<File>> callback;
+    private Set<File> watchedFiles;
+    // Fields set in start()
+    private ListeningScheduledExecutorService executorService;
+    private ListenableScheduledFuture<?> future;
+    // Cache of last known file state, updated at every polling cycle
+    private final AtomicReference<Map<File, FileMetadata>> metadataCacheRef;
+
+    private static final Logger log = Logger.get(PollingMultiFileWatcher.class);
+
+    /**
+     * Creates a new watcher. The watcher doesn't start scanning files on disk until the {@code start()} method is
+     * called.
+     *
+     * @param initialDelay how long to wait until the first scan of the files.
+     * @param interval how often to rescan the files.
+     * @param timeUnit time unit for {@code initialDelay} and {@code interval}.
+     */
+    public PollingMultiFileWatcher(long initialDelay, long interval, TimeUnit timeUnit) {
+        this.initialDelay = initialDelay;
+        this.interval = interval;
+        this.timeUnit = timeUnit;
+        this.executorService = null;
+        this.future = null;
+        this.metadataCacheRef = new AtomicReference<>(ImmutableMap.of());
+        watchedFiles = ImmutableSet.of();
+        callback = null;
+    }
+
+    /**
+     * Starts polling the watched files for changes.
+     */
+    public void start(Set<File> files, Consumer<Set<File>> callback) {
+        if (isStarted()) {
+            throw new IllegalStateException("start() should not be called more than once");
+        }
+        Preconditions.checkArgument(files.size() > 0, "must specify at least 1 file to watch for changes");
+        this.callback = requireNonNull(callback);
+        watchedFiles = ImmutableSet.copyOf(files);
```

Comment:
slightly better form i guess to copy files first, then check size of the copy > 0?